lib/elixir/lib/module/parallel_checker.ex

```diff
@@ -317,6 +317,7 @@ defmodule Module.ParallelChecker do
 
   defp cache_from_chunk(ets, module) do
     with {^module, binary, _filename} <- :code.get_object_code(module),
+         {:ok, ^module} <- binary |> :beam_lib.info() |> Keyword.fetch(:module),
```

Comment:
I think in this case we check it matches upstream, because the module was previously defined in a previous compiler pass:

```suggestion
```

Reply:
Actually, I think you added a test showing this is necessary for lazily loaded modules. :)